fails to build on 32 bit

The build on 32 bit archlinux fails due to some type cast from `size_t *` to `long unsigned int *`:

```
+ exec meson setup --prefix /usr --libdir /usr/lib --libexecdir /usr/lib --bindir /usr/bin --sbindir /usr/bin --includedir /usr/include --datadir /usr/share --mandir /usr/share/man --infodir /usr/share/info --localedir /usr/share/locale --sysconfdir /etc --localstatedir /var --sharedstatedir /var/lib --buildtype release --auto-features enabled --wrap-mode nofallback -D b_lto=true build grim
The Meson build system
Version: 0.48.2
Source dir: /build/grim/src/grim
Build dir: /build/grim/src/build
Build type: native build
Project name: grim
Project version: 0.0.0
Appending CFLAGS from environment: '-march=i686 -mtune=generic -O2 -pipe -fstack-protector-strong -fno-plt'
Appending LDFLAGS from environment: '-Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now'
Appending CPPFLAGS from environment: '-D_FORTIFY_SOURCE=2'
Native C compiler: cc (gcc 8.2.1 "cc (GCC) 8.2.1 20180831")
Build machine cpu family: x86
Build machine cpu: i686
Found pkg-config: /usr/bin/pkg-config (1.5.4)
Dependency cairo found: YES 1.16.0
Dependency wayland-client found: YES 1.16.0
Dependency wayland-protocols found: YES 1.16
Dependency libjpeg found: YES 2.0.0
Library m found: YES
Program wayland-scanner found: YES (/usr/bin/wayland-scanner)
Program scdoc found: NO
Build targets in project: 2
Option werror is: True [default: true]
Found ninja-1.8.2 at /usr/bin/ninja
ninja: Entering directory `build'
[1/16] Generating 'protocol/protocol@@client_protos@sta/wlr-screencopy-unstable-v1-protocol.c'.
[2/16] Generating 'protocol/protocol@@client_protos@sta/xdg-output-unstable-v1-protocol.c'.
[3/16] Generating 'protocol/protocol@@client_protos@sta/xdg-output-unstable-v1-client-protocol.h'.
[4/16] Generating 'protocol/protocol@@client_protos@sta/wlr-screencopy-unstable-v1-client-protocol.h'.
[5/16] Generating 'grim@exe/wlr-screencopy-unstable-v1-client-protocol.h'.
[6/16] Generating 'grim@exe/xdg-output-unstable-v1-client-protocol.h'.
[7/16] Compiling C object 'protocol/protocol@@client_protos@sta/meson-generated_wlr-screencopy-unstable-v1-protocol.c.o'.
[8/16] Compiling C object 'protocol/protocol@@client_protos@sta/meson-generated_xdg-output-unstable-v1-protocol.c.o'.
[9/16] Linking static target protocol/libclient_protos.a.
[10/16] Compiling C object 'grim@exe/box.c.o'.
[11/16] Compiling C object 'grim@exe/cairo_jpg.c.o'.
FAILED: grim@exe/cairo_jpg.c.o 
cc -Igrim@exe -I. -I../grim -I../grim/include -I/usr/include/cairo -I/usr/include/glib-2.0 -I/usr/lib/glib-2.0/include -I/usr/lib/libffi-3.2.1/include -I/usr/include/pixman-1 -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/harfbuzz -I/usr/include/uuid -flto -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -Werror -std=c11 -O3 -Wno-unused-parameter -march=i686 -mtune=generic -O2 -fstack-protector-strong -fno-plt -D_FORTIFY_SOURCE=2  -MD -MQ 'grim@exe/cairo_jpg.c.o' -MF 'grim@exe/cairo_jpg.c.o.d' -o 'grim@exe/cairo_jpg.c.o' -c ../grim/cairo_jpg.c
../grim/cairo_jpg.c: In function ‘cairo_surface_write_to_jpeg_mem’:
../grim/cairo_jpg.c:50:30: error: passing argument 3 of ‘jpeg_mem_dest’ from incompatible pointer type [-Werror=incompatible-pointer-types]
  jpeg_mem_dest(&cinfo, data, len);
                              ^~~
In file included from ../grim/cairo_jpg.c:13:
/usr/include/jpeglib.h:922:43: note: expected ‘long unsigned int *’ but argument is of type ‘size_t *’ {aka ‘unsigned int *’}
                            unsigned long *outsize);
                            ~~~~~~~~~~~~~~~^~~~~~~
../grim/cairo_jpg.c: In function ‘cj_write’:
../grim/cairo_jpg.c:83:42: error: comparison of integer expressions of different signedness: ‘ssize_t’ {aka ‘int’} and ‘unsigned int’ [-Werror=sign-compare]
  if (write((long) closure, data, length) < length) {
                                          ^
cc1: all warnings being treated as errors
[12/16] Compiling C object 'grim@exe/buffer.c.o'.
[13/16] Compiling C object 'grim@exe/main.c.o'.
ninja: build stopped: subcommand failed.
```

regards,
deep-42-thought

--- a/cairo_jpg.c
+++ b/cairo_jpg.c
@@ -7,7 +7,9 @@
 #include <stdlib.h>
 #include <sys/types.h>
 #include <sys/stat.h>
+#include <assert.h>
 #include <fcntl.h>
+#include <limits.h>
 #include <unistd.h>
 #include <cairo.h>
 #include <jpeglib.h>
@@ -15,7 +17,7 @@
 #include "cairo_jpg.h"
 
 cairo_status_t cairo_surface_write_to_jpeg_mem(cairo_surface_t *sfc,
-		unsigned char **data, size_t *len, int quality) {
+		unsigned char **data, unsigned long *len, int quality) {
 	struct jpeg_compress_struct cinfo;
 	struct jpeg_error_mgr jerr;
 	JSAMPROW row_pointer[1];
@@ -91,18 +93,17 @@ cairo_status_t cairo_surface_write_to_jpeg_stream(cairo_surface_t *sfc,
 		cairo_write_func_t write_func, void *closure, int quality) {
 	cairo_status_t e;
 	unsigned char *data = NULL;
-	size_t len = 0;
+	unsigned long len = 0;
 
 	e = cairo_surface_write_to_jpeg_mem(sfc, &data, &len, quality);
-	if (e != CAIRO_STATUS_SUCCESS) {
-		return e;
+	if (e == CAIRO_STATUS_SUCCESS) {
+		assert(sizeof(unsigned long) <= sizeof(size_t)
+			|| !(len >> (sizeof(size_t) * CHAR_BIT)));
+		e = write_func(closure, data, len);
+		free(data);
 	}
 
-	e = write_func(closure, data, len);
-
-	free(data);
 	return e;
-
 }
 
 cairo_status_t cairo_surface_write_to_jpeg(cairo_surface_t *sfc,

--- a/include/cairo_jpg.h
+++ b/include/cairo_jpg.h
@@ -3,7 +3,7 @@
 
 #include <cairo.h>
 
-cairo_status_t cairo_surface_write_to_jpeg_mem(cairo_surface_t *sfc, unsigned char **data, size_t *len, int quality);
+cairo_status_t cairo_surface_write_to_jpeg_mem(cairo_surface_t *sfc, unsigned char **data, unsigned long *len, int quality);
 cairo_status_t cairo_surface_write_to_jpeg_stream(cairo_surface_t *sfc, cairo_write_func_t write_func, void *closure, int quality);
 cairo_status_t cairo_surface_write_to_jpeg(cairo_surface_t *sfc, const char *filename, int quality);
 

